src/Lean/Data/Lsp/Utf16.lean

```diff
@@ -20,93 +21,63 @@ end Char
 
 namespace String
 
-/-- LSP is EOL agnostic, so we first split on "\r\n" to avoid possibly redundant line breaks. -/
-def splitOnEOLs (s : String) : List String := do
-line ← s.splitOn "\r\n";
-line ← line.splitOn "\n";
-line ← line.splitOn "\r";
-pure line
-
 private def csize (c : Char) : Nat :=
 c.utf16Size.toNat
 
-private def codepointPosToUtf16PosAux (s : String) : Nat → Pos → Nat → Nat
+def utf16Length (s : String) : Nat :=
+s.foldr (fun c acc => csize c + acc) 0
+
+private def codepointPosToUtf16PosFromAux (s : String) : Nat → Pos → Nat → Nat
 | 0,    utf8pos, utf16pos => utf16pos
-| cp+1, utf8pos, utf16pos => codepointPosToUtf16PosAux cp (s.next utf8pos) (utf16pos + csize (s.get utf8pos))
+| cp+1, utf8pos, utf16pos => codepointPosToUtf16PosFromAux cp (s.next utf8pos) (utf16pos + csize (s.get utf8pos))
+
+/-- Computes the UTF-16 offset of the `n`-th Unicode codepoint
+in the substring of `s` starting at UTF-8 offset `off`.
+Yes, this is actually useful.-/
```

Comment:
:smile:

Reply:
We *could* make these APIs nicer by packing `s` and `off` into a `String.Iterator`, at which point you wouldn't have to mention the UTF-8 offset anymore. But it doesn't seem to buy us anything here, so let's leave it as is.